### PR TITLE
refactor outbound/hmail.process_ehlo_data()

### LINKED ISSUE #2483: outbound STARTTLS  not functional
### system info
Haraka |
 --- | :---
Node | v8.11.3
OS | Ubuntu SMP Fri May 25 21:47:34 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.0.2g  1 Mar 2016

### Expected behavior
STARTTLS being used for outbound email  to Gmail and Gmail has no "no encryption" alert.

### Observed behavior
I've tls.ini configured with key= and cert= correctly.  How ever the outbound delivery queue is not respecting the remote server's STARTTLS command in EHLO response. 
That result in the outbound email is sent to the remote server with no TLS encryption.
Gmail is currently indicating a red mark for this kind of emails which confuses the receiver. 

#### The console output:
```
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-mx.google.com at your service, [54.169.47.2]\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-SIZE 157286400\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-8BITMIME\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-STARTTLS\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-ENHANCEDSTATUSCODES\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-PIPELINING\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250-CHUNKING\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250 SMTPUTF8\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] C: MAIL FROM:<pauking@amail.sinoqual.com>
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250 2.1.0 OK q61-v6si10368359plb.93 - gsmtp\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] C: RCPT TO:<sunnipaul@gmail.com>
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250 2.1.5 OK q61-v6si10368359plb.93 - gsmtp\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] C: DATA
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 354  Go ahead q61-v6si10368359plb.93 - gsmtp\r\n
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] C: .
[PROTOCOL] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound] S: 250 2.0.0 OK 1532426523 q61-v6si10368359plb.93 - gsmtp\r\n
[NOTICE] [D2E02A57-9674-45C9-8C87-4EAFD3BA48D9.1.1] [outbound]  delivered file=1532426521449_1532426521449_0_18255_dGm6jl_1_ip-172-31-20-173 domain=gmail.com host=gmail-smtp-in.l.google.com ip=172.217.194.26 port=25 mode=SMTP tls=N auth=N response="OK 1532426523 q61-v6si10368359plb.93 - gsmtp" delay=2.015 fails=0 rcpts=1/0/0
```
image
However, if I change   ```at line 556 in hmail.js``` as following: 

 > if (obtls.cfg !== undefined && ~~obtls.cfg.no_tls_hosts !== undefined &&~~ 

everything works correctly.

### Steps to reproduce
1. config tls.ini with proper key and cert  and ``` enable_tls = true ``` , setup mx record for the sending machine and run haraka
2. send email with a client  to gmail 
3. gmail indicate no encryption is used.

### Attached python script to send email 
``` 
to = 'someone@gmail.com'
from_addr = 'someone@somedomain.com'
smtpserver = smtplib.SMTP("somedomain.com",25)
smtpserver.ehlo()
smtpserver.starttls()
smtpserver.ehlo
smtpserver.login('someone', 'somepassword')
header = 'To:' + to + '\n' + 'From: ' + from_addr + '\n' + 'Subject:testing \n'
print header
msg = header + '\n this is test msg \n\n'
smtpserver.sendmail(from_addr, to, msg)
print 'done!'
smtpserver.close()
```

### PR DESCRIPTION
**Pull request by msimerson**
Fixes #2483

Changes proposed in this pull request:
- refactor outbound/hmail.process_ehlo_data

### CODE REVIEW

File: outbound/hmail.js [hunk 1]
```diff
@@ -548,37 +548,43 @@ class HMailItem extends events.EventEmitter {
...
 
+            if (obtls.cfg === undefined) return auth_and_mail_phase();  // no outbound TLS config
+
+            // TLS is configured and available
+
+            // net-utils before 1.1.1 would throw if sent an empty list
+            if (obtls.cfg.no_tls_hosts === undefined) obtls.cfg.no_tls_hosts = [];
+
+            // TLS exclude lists checks for MX host or remote domain
+            if (net_utils.ip_in_list(obtls.cfg.no_tls_hosts, host) return auth_and_mail_phase();
```
Comment (baudehlo) [inline outbound/hmail.js]: Missing bracket

File: outbound/hmail.js [hunk 2]
```diff
@@ -548,37 +548,43 @@ class HMailItem extends events.EventEmitter {
...
-            // TLS
-            if (!secured && smtp_properties.tls && cfg.enable_tls) {
-                if (obtls.cfg !== undefined && obtls.cfg.no_tls_hosts !== undefined &&
-                    !net_utils.ip_in_list(obtls.cfg.no_tls_hosts, host) &&
-                    !net_utils.ip_in_list(obtls.cfg.no_tls_hosts, self.todo.domain)) {
+            if (secured) return auth_and_mail_phase();              // TLS already negotiated
+            if (!cfg.enable_tls) return auth_and_mail_phase();      // TLS not configured
+            if (!smtp_properties.tls) return auth_and_mail_phase(); // TLS not advertised by remote
 
+            if (obtls.cfg === undefined) return auth_and_mail_phase();  // no outbound TLS config
```
[the remark this answers is not in the document]
Comment (msimerson) [reply outbound/hmail.js]: It's possible to get here before the config file is loaded. That was the root cause of one of the crashing error issues posted recently.

**Comment by baudehlo** [comment]
Yikes. If that's the case we should probably tempfail though, surely?

File: outbound/hmail.js [hunk 3]
```diff
@@ -548,37 +548,43 @@ class HMailItem extends events.EventEmitter {
...
+
+            // TLS is configured and available
+
+            // net-utils before 1.1.1 would throw if sent an empty list
+            if (obtls.cfg.no_tls_hosts === undefined) obtls.cfg.no_tls_hosts = [];
```
[the remark this answers is not in the document]
Comment (msimerson) [reply outbound/hmail.js]: agreed

**Comment by baudehlo** [comment]
In theory you might just be able to do `return hmail.temp_fail(reason)`.
Though I wonder if it will eventually hit finished_processing_mail() too if
you do that.

**Comment by msimerson** [comment]
IIRC, in the cases where this code runs before the tls config is loaded, it's immediately after Haraka is started while messages in the outbound queue are ready to go. In such a case, the message would get delivered without TLS, even if TLS was configured and would have been available in another second or two. The consequence of tempfail would be deferring the message delivery by N minutes?

**Comment by baudehlo** [comment]
That shouldn't happen any more due to this:
https://github.com/haraka/Haraka/blob/master/outbound/queue.js#L74

**Comment by msimerson** [comment]
If that's the case, then that condition will never be true and this is a non-issue. What do you think about adding something like the following, and then inspecting for those errors later?  (I don't have time to test a `return hmail.temp_fail('TLS config not loaded yet');` right now and I'd like to be confident it'll do something intelligent before I add it).

```diff
-            if (obtls.cfg === undefined) return auth_and_mail_phase();  // no outbound TLS config
+            if (obtls.cfg === undefined) {
+                self.logerror(`Oops, TLS config not loaded yet!`);
+                return auth_and_mail_phase();  // no outbound TLS config
+            }
```

**Comment by baudehlo** [comment]
Are you going to make it a logcrit? Or do you think it should remain
logerror?

**Comment by msimerson** [comment]
I generally grep for ERROR in the logs and expect to find errors there. 

To me, a CRIT means there’s been a severe application error that requires the daemon to shut down. As far as I can tell, the entire reason for that check is just an async/timing issue. That’s a long ways away from a CRIT IMO.

**Comment by baudehlo** [comment]
Fair argument. +1

**Pull request merged**